roachtest: sqlsmith unable to encode JSON table key

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2503575&tab=buildLog) on [release-20.2@89dd2d14c29787378c434ed54937757ef5d9877c](https://github.com/cockroachdb/cockroach/commits/89dd2d14c29787378c434ed54937757ef5d9877c):

```
The test failed on branch=release-20.2, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:181,sqlsmith.go:211,test_runner.go:755: error: pq: internal error: unexpected error from the vectorized engine: unable to encode table key: *tree.DJSON
		stmt:
		WITH
			with_3378 (col_19839)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									('false':::JSONB, '1970-01-05 16:57:40.000665+00:00':::TIMESTAMPTZ, ARRAY[(-29788):::INT8,21647:::INT8,(-29552):::INT8,(-8008):::INT8], 'BOX(-1.011475312307369 -2.019039666757337,-0.6528405112126936 -0.8061805935915762)':::BOX2D)
								)
						)
							AS tab_8477 (col_19839)
					EXCEPT ALL
						SELECT
							*
						FROM
							(
								VALUES
									(
										('[[[true], [], {}, "b", {}], {"a": []}, {"c": 2.05750813403415}]':::JSONB, '1970-01-10 05:23:26.000428+00:00':::TIMESTAMPTZ, ARRAY[(-31919):::INT8], 'BOX(0.5953204032504589 -0.10129389723789606,1.1109876367775304 0.4610822966768364)':::BOX2D)
									),
									(
										('{"test": "json"}':::JSONB, '1970-01-01 18:58:12.000671+00:00':::TIMESTAMPTZ, ARRAY[24506:::INT8,(-16669):::INT8,7580:::INT8], 'BOX(0.29452191917172077 -0.21929511336462593,1.2697864663294478 0.8070809385447182)':::BOX2D)
									),
									(
										('{}':::JSONB, '1970-01-09 17:56:40.000016+00:00':::TIMESTAMPTZ, ARRAY[11362:::INT8,(-11037):::INT8,18298:::INT8], 'BOX(-0.6271419682324558 -0.8910290618767774,0.5682509653436363 0.8149505896862345)':::BOX2D)
									),
									(
										('[[{"mn8M|": "{]PUd"}], {"bar": [[[[], [[], "foo", []]]], {"a": 1.0830811802633054}]}, []]':::JSONB, '1970-01-11 00:44:51.000992+00:00':::TIMESTAMPTZ, ARRAY[]:::INT2[], 'BOX(-0.3890135712093068 -1.8282316755909567,1.480562509603321 -0.39320967028393294)':::BOX2D)
									)
							)
								AS tab_8478 (col_19840)
				),
			with_3379 (col_19841) AS (SELECT * FROM (VALUES (0:::OID)) AS tab_8479 (col_19841))
		SELECT
			tab_8480.col1_6 AS col_19842,
			tab_8480.col1_3 AS col_19843,
			tab_8480.tableoid AS col_19844,
			timezone(tab_8480.col1_5::STRING, tab_8480.col1_1::TIMETZ)::TIMETZ AS col_19845
		FROM
			with_3378 AS cte_ref_969, defaultdb.public.table1@[0] AS tab_8480;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2503575&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/col/coldataext/datum_vec.go
+++ b/pkg/col/coldataext/datum_vec.go
@@ -87,7 +87,10 @@ func (d *Datum) Hash(da *rowenc.DatumAlloc) []byte {
 	// decode ed for fingerprinting, so we pass in nil memory account.
 	b, err := ed.Fingerprint(context.TODO(), d.ResolvedType(), da, nil /* appendTo */, nil /* acc */)
 	if err != nil {
-		colexecerror.InternalError(err)
+		// Since we know that the memory accounting error cannot occur in
+		// Fingerprint, we only can get an expected error (e.g. unable to encode
+		// JSON as a key), so we propagate all of them accordingly.
+		colexecerror.ExpectedError(err)
 	}
 	return b
 }

--- a/pkg/sql/logictest/testdata/logic_test/hash_join
+++ b/pkg/sql/logictest/testdata/logic_test/hash_join
@@ -216,3 +216,10 @@ SELECT * FROM t44797_2 WHERE EXISTS (SELECT * FROM t44797_2 AS l, t44797_3 AS r 
 ----
 1
 2
+
+# Regression test for incorrectly propagating an error in the vectorized engine.
+statement ok
+CREATE TABLE table57696(col_table TIME NOT NULL)
+
+statement error unable to encode JSON as a table key\nHINT:.*\n.*35706.*
+WITH cte (col_cte) AS ( SELECT * FROM ( VALUES ( ( 'false':::JSONB, '1970-01-05 16:57:40.000665+00:00':::TIMESTAMPTZ ) ) ) EXCEPT ALL SELECT * FROM ( VALUES ( ( ' [ [[true], [], {}, "b", {}], {"a": []}, {"c": 2.05750813403415} ] ':::JSONB, '1970-01-10 05:23:26.000428+00:00':::TIMESTAMPTZ ) ) ) ) SELECT * FROM cte, table57696

--- a/pkg/sql/rowenc/BUILD.bazel
+++ b/pkg/sql/rowenc/BUILD.bazel
@@ -31,6 +31,7 @@ go_library(
         "//pkg/util/bitarray",
         "//pkg/util/duration",
         "//pkg/util/encoding",
+        "//pkg/util/errorutil/unimplemented",
         "//pkg/util/ipaddr",
         "//pkg/util/json",
         "//pkg/util/log",

--- a/pkg/sql/rowenc/column_type_encoding.go
+++ b/pkg/sql/rowenc/column_type_encoding.go
@@ -22,6 +22,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/bitarray"
 	"github.com/cockroachdb/cockroach/pkg/util/duration"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 	"github.com/cockroachdb/cockroach/pkg/util/ipaddr"
 	"github.com/cockroachdb/cockroach/pkg/util/json"
 	"github.com/cockroachdb/cockroach/pkg/util/timetz"
@@ -183,6 +184,8 @@ func EncodeTableKey(b []byte, val tree.Datum, dir encoding.Direction) ([]byte, e
 			return encoding.EncodeBytesAscending(b, t.PhysicalRep), nil
 		}
 		return encoding.EncodeBytesDescending(b, t.PhysicalRep), nil
+	case *tree.DJSON:
+		return nil, unimplemented.NewWithIssue(35706, "unable to encode JSON as a table key")
 	}
 	return nil, errors.Errorf("unable to encode table key: %T", val)
 }

